http3/frames.go

```diff
@@ -194,3 +195,41 @@ func (f *settingsFrame) Append(b []byte) []byte {
 	}
 	return b
 }
+
+type goawayFrame struct {
+	ID quic.StreamID
+}
+
+// max length of an encoded quic varint: quicvarint.Len(quicvarint.Max)
+const maxQuicVarintLen = 8
+
+func parseGoawayFrame(r io.Reader, l uint64) (*goawayFrame, error) {
+	if l > maxQuicVarintLen {
```

Comment:
Easier: just parse the varint, and check that the length matches the number of bytes parsed afterwards.

Reply:
There is a check afterwards to see if data read matches the varint it encodes. This check is necessary.

Reply:
No, it is not. This code is just unnecessarily complicated. Use `quicvarint.Read`.

Reply:
A malformed goaway frame can have an arbitrary large length defined. And `quicvarint.Read` is used later.

Reply:
Just parse the two varints, then check that the length is consistent. Saves you 10 LOC and multiple allocs.

Reply:
I used an `io.LimitedReader` and `bufio.Reader` to read from the frame body. Do you think it's ok now?

Reply:
Still too complicated.

1. Read varint (length)
2. Read varint (length)
3. Check that the bytes consumed by (2) match the value read in (1).

Reply:
You have to wrap the reader with a `bufio.Reader` to be able to use Read varint. And to limit the number of bytes read a limited reader is needed, or else the stream may be corrupt due to more data read than needed.

Reply:
Any feedbacks? @marten-seemann

Reply:
My previous comment still stands.

Reply:
Show me some code snippets. I don't think I know how to think of this. Double read?

Reply:
Sure. I'll do this once the rest of the PR is in a shape where we could merge it.